drivers/timer/nrf_rtc_timer.c

```diff
@@ -523,87 +86,73 @@ int z_clock_driver_init(struct device *device)
 
 	clock_control_on(clock, (void *)CLOCK_CONTROL_NRF5_K32SRC);
 
-	rtc_past = 0;
-
-#ifdef CONFIG_TICKLESS_IDLE
-	expected_sys_ticks = 1;
-#endif /* CONFIG_TICKLESS_IDLE */
-
 	/* TODO: replace with counter driver to access RTC */
-	SYS_CLOCK_RTC->PRESCALER = 0;
-	nrf_rtc_cc_set(SYS_CLOCK_RTC, RTC_CC_IDX,
-		       sys_clock_hw_cycles_per_tick());
-	nrf_rtc_event_enable(SYS_CLOCK_RTC, RTC_EVTENSET_COMPARE0_Msk);
-	nrf_rtc_int_enable(SYS_CLOCK_RTC, RTC_INTENSET_COMPARE0_Msk);
+	RTC->PRESCALER = 0;
+	nrf_rtc_cc_set(RTC, 0, CYC_PER_TICK);
+	nrf_rtc_event_enable(RTC, RTC_EVTENSET_COMPARE0_Msk);
+	nrf_rtc_int_enable(RTC, RTC_INTENSET_COMPARE0_Msk);
 
 	/* Clear the event flag and possible pending interrupt */
-	RTC_CC_EVENT = 0;
+	RTC->EVENTS_COMPARE[0] = 0;
 	NVIC_ClearPendingIRQ(NRF5_IRQ_RTC1_IRQn);
 
 	IRQ_CONNECT(NRF5_IRQ_RTC1_IRQn, 1, rtc1_nrf5_isr, 0, 0);
 	irq_enable(NRF5_IRQ_RTC1_IRQn);
 
-	nrf_rtc_task_trigger(SYS_CLOCK_RTC, NRF_RTC_TASK_CLEAR);
-	nrf_rtc_task_trigger(SYS_CLOCK_RTC, NRF_RTC_TASK_START);
+	nrf_rtc_task_trigger(RTC, NRF_RTC_TASK_CLEAR);
+	nrf_rtc_task_trigger(RTC, NRF_RTC_TASK_START);
+
+	if (!IS_ENABLED(TICKLESS_KERNEL)) {
+		set_comparator(now() + CYC_PER_TICK);
+	}
 
 	return 0;
 }
 
-u32_t _timer_cycle_get_32(void)
+void z_clock_set_timeout(s32_t ticks, bool idle)
 {
-	u32_t ticked_cycles;
-	u32_t elapsed_cycles;
-
-	/* Number of timer cycles announced as ticks so far. */
-	ticked_cycles = z_tick_get() * sys_clock_hw_cycles_per_tick();
-
-	/* Make sure that compiler will not reverse access to RTC and
-	 * z_tick_get().
-	 */
-	compiler_barrier();
-
-	/* Number of timer cycles since last announced tick we know about.
-	 *
-	 * The value of RTC_COUNTER is not reset on tick, so it will
-	 * compensate potentialy missed update of z_tick_get()
-	 * which could have happen between the ticked_cycles calculation
-	 * and the code below.
-	 */
-	elapsed_cycles = (RTC_COUNTER - ticked_cycles) & RTC_MASK;
-
-	return ticked_cycles + elapsed_cycles;
-}
+	ARG_UNUSED(idle);
 
-#ifdef CONFIG_SYSTEM_CLOCK_DISABLE
-/**
- *
- * @brief Stop announcing sys ticks into the kernel
- *
- * This routine disables the RTC1 so that timer interrupts are no
- * longer delivered.
- *
- * @return N/A
- */
-void sys_clock_disable(void)
-{
-	unsigned int key;
+#ifdef CONFIG_TICKLESS_KERNEL
+	ticks = ticks == K_FOREVER ? MAX_TICKS : ticks;
+	ticks = max(min(ticks - 1, (s32_t)MAX_TICKS), 0);
 
-	key = irq_lock();
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	u32_t cyc, t = now();
 
-	irq_disable(NRF5_IRQ_RTC1_IRQn);
+	/* Round up to next tick boundary */
+	cyc = ticks * CYC_PER_TICK + counter_sub(t, last_count);
```

Comment:
What about potential overflow in these calculations?

Reply:
Ticks gets clamped to [0,MAX_TICKS] in the code above this spot, which is calibrated to prevent that.

Reply:
To be clear: that's about preventing overflow beyond 24 bits of cycles in a single timeout.

There's no need to prevent overflow in the calculation of the next timeout value itself, the cycle counter is cyclic and just rolls over.